src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -163,122 +158,69 @@ CircularBoundedJournal::open_device_read_header(rbm_abs_addr start)
-{
-  LOG_PREFIX(CircularBoundedJournal::append_record);
-  std::vector<std::pair<rbm_abs_addr, bufferlist>> writes;
-  if (addr + bl.length() <= get_journal_end()) {
-    writes.push_back(std::make_pair(addr, bl));
-  } else {
-    // write remaining data---in this case,
-    // data is splited into two parts before due to the end of CircularBoundedJournal.
-    // the following code is to write the second part
-    bufferlist first_write, next_write;
-    first_write.substr_of(bl, 0, get_journal_end() - addr);
-    writes.push_back(std::make_pair(addr, first_write));
-    next_write.substr_of(
-      bl, first_write.length(), bl.length() - first_write.length());
-    writes.push_back(std::make_pair(get_start_addr(), next_write));
-  }
-
-  return seastar::do_with(
-    std::move(bl),
-    [this, writes=move(writes), FNAME](auto& bl) mutable
-  {
-    DEBUG("original bl length {}", bl.length());
-    return write_ertr::parallel_for_each(
-      writes,
-      [this, FNAME](auto& p) mutable
-    {
-      DEBUG(
-	"append_record: offset {}, length {}",
-	p.first,
-	p.second.length());
-      return device_write_bl(p.first, p.second 
-      ).handle_error(
-	write_ertr::pass_further{},
-	crimson::ct_error::assert_all{ "Invalid error device->write" }
-      ).safe_then([]() {
-	return write_ertr::now();
-      });
-    });
-  });
-}
-
 CircularBoundedJournal::submit_record_ret CircularBoundedJournal::submit_record(
   record_t &&record,
   OrderingHandle &handle)
 {
   LOG_PREFIX(CircularBoundedJournal::submit_record);
   assert(write_pipeline);
+  assert(circulation_seq != NULL_SEG_SEQ);
   auto r_size = record_group_size_t(record.size, get_block_size());
   auto encoded_size = r_size.get_encoded_length();
-  if (get_written_to() +
-      ceph::encoded_sizeof_bounded<record_group_header_t>() > get_journal_end()) {
-    // not enough space between written_to and the end of journal,
-    // so that set written_to to the beginning of cbjournal to append 
-    // the record at the start address of cbjournal
-    // |        cbjournal      |
-    // 	          v            v
-    //      written_to <-> the end of journal
-    set_written_to(get_start_addr());
-  }
   if (encoded_size > get_available_size()) {
-    ERROR(
-      "CircularBoundedJournal::submit_record: record size {}, but available size {}",
-      encoded_size,
-      get_available_size()
-      );
+    ERROR("record size {}, but available size {}",
+          encoded_size, get_available_size());
     return crimson::ct_error::erange::make();
   }
+  if (encoded_size + get_written_to() > get_journal_end()) {
+    DEBUG("roll");
+    set_written_to(get_start_addr());
+    ++circulation_seq;
+    if (encoded_size > get_available_size()) {
+      ERROR("rolled, record size {}, but available size {}",
+            encoded_size, get_available_size());
+      return crimson::ct_error::erange::make();
+    }
+  }
 
   journal_seq_t j_seq {
-    cur_segment_seq++,
+    circulation_seq,
```

Comment:
Could you please explain the purpose of circulation_seq? Sequence number here is originally used to check that records are stored in order correctly---it was increased every record.

Reply:
In the segmented-journal, segment-seq only increase during rolling to a new segment, which is very similar to rolling to the start of the circular journal, it won't increment for every record, see https://github.com/ceph/ceph/blob/main/src/crimson/os/seastore/journal/segment_allocator.cc#L38-L39

Instead, `journal_seq_t` is the actual thing that will be increased by every record as it consists of [segment-seq, offset].